File: fractal-gas-web/web/lab/agent-types.js
// Scene archetypes are portable data, shared by native physics and rendering.
// Physics overlays are shallow; shape arrays replace inherited arrays in full.
export function resolveAgentTypes(definitions = {}) {
  if (
    !definitions ||
    Array.isArray(definitions) ||
    typeof definitions !== "object"
  )
    throw new Error("agent_types must be an object");
  const resolved = new Map(),
    visiting = new Set();
  if (Object.keys(definitions).length > 256)
    throw new Error("At most 256 agent types are supported");
  const resolve = (name) => {
    if (resolved.has(name)) return resolved.get(name);
    if (!Object.hasOwn(definitions, name))
      throw new Error(`Unknown agent type: ${name}`);
    if (visiting.has(name))
      throw new Error(`Cyclic agent type inheritance: ${name}`);
    const def = definitions[name];
    if (!name || !def || Array.isArray(def) || typeof def !== "object")
      throw new Error("Agent types need a name and an object definition");
    visiting.add(name);
    const parent = def.extends == null ? {} : resolve(def.extends);
    for (const field of ["physics", "visual"])
      if (
        def[field] != null &&
        (typeof def[field] !== "object" || Array.isArray(def[field]))
      )
        throw new Error(`Agent ${field} defaults must be an object`);
    if (def.physics && Object.hasOwn(def.physics, "agent_type"))
      throw new Error("Use extends to inherit agent types");
    const type = {
      ...parent,
      ...def,
      physics: { ...parent.physics, ...def.physics },
      visual: { ...parent.visual, ...def.visual },
    };
    visiting.delete(name);
    resolved.set(name, type);
    return type;
  };
  Object.keys(definitions).forEach(resolve);
  return resolved;
}
export function resolveBodies(scene) {
  const types = resolveAgentTypes(scene.agent_types);
  return scene.bodies.map((body) => {
    if (body.agent_type == null) return { ...body };
    const type = types.get(body.agent_type);
    if (!type) throw new Error(`Unknown agent type: ${body.agent_type}`);
    return {
      ...type.physics,
      ...body,
      visual: { ...type.visual, ...body.visual },
    };
  });
}

File: fractal-gas-web/src/control/agent_types.cpp
#include "control/agent_types.hpp"

#include <algorithm>

namespace fg::control {
namespace {
void overlay(Json& target, const Json& source) {
  if (source.kind != Json::Null && source.kind != Json::Object)
    throw std::invalid_argument("Agent physics defaults must be an object");
  target.kind = Json::Object;
  for (const auto& [key, value] : source.object) {
    if (key == "agent_type")
      throw std::invalid_argument("Use extends to inherit agent types");
    target.object[key] = value;
  }
}
}  // namespace
AgentTypes::AgentTypes(const Json& definitions) {
  if (definitions.kind != Json::Null && definitions.kind != Json::Object)
    throw std::invalid_argument("agent_types must be an object");
  if (definitions.object.size() > 256)
    throw std::invalid_argument("At most 256 agent types are supported");
  definitions_ = definitions.object;
  std::vector<std::string> visiting;
  for (const auto& [name, definition] : definitions_) resolve(name, visiting);
}
Json AgentTypes::resolve(const std::string& name,
                         std::vector<std::string>& visiting) {
  if (resolved_.count(name)) return resolved_.at(name);
  auto found = definitions_.find(name);
  if (found == definitions_.end())
    throw std::invalid_argument("Unknown agent type: " + name);
  if (std::find(visiting.begin(), visiting.end(), name) != visiting.end())
    throw std::invalid_argument("Cyclic agent type inheritance: " + name);
  const Json& definition = found->second;
  if (name.empty() || definition.kind != Json::Object)
    throw std::invalid_argument(
        "Agent types need a name and an object definition");
  visiting.push_back(name);
  Json defaults;
  if (definition["extends"].kind != Json::Null)
    defaults = resolve(definition["extends"].str(), visiting);
  overlay(defaults, definition["physics"]);
  visiting.pop_back();
  return resolved_[name] = std::move(defaults);
}
Json AgentTypes::body(const Json& instance) const {
  if (instance.kind != Json::Object)
    throw std::invalid_argument("Body must be an object");
  if (instance["agent_type"].kind == Json::Null) return instance;
  const std::string name = instance["agent_type"].str();
  auto found = resolved_.find(name);
  if (found == resolved_.end())
    throw std::invalid_argument("Unknown agent type: " + name);
  Json result = found->second;
  for (const auto& [key, value] : instance.object) result.object[key] = value;
  return result;
}
}  // namespace fg::control
